pkg/auth/authproviders/oidc/backend_impl.go

```diff
@@ -630,9 +649,66 @@ func userInfoToExternalClaims(userInfo *userInfoType) *tokens.ExternalUserClaim
 	return claim
 }
 
-func (p *backendImpl) injectHTTPClient(ctx context.Context) context.Context {
-	if p.httpClient != nil {
-		return oidc.ClientContext(ctx, p.httpClient)
+type claimExtractor interface {
+	Claims(v interface{}) error
+}
+
+func extractCustomClaims(externalUserClaim *tokens.ExternalUserClaim, mappings map[string]string, claimExtractor claimExtractor) error {
+	claims := make(map[string]interface{}, 0)
+	if err := claimExtractor.Claims(&claims); err != nil {
+		return errors.Wrap(err, "failed to extract claims")
 	}
-	return ctx
+	for fromClaimPath, toClaimName := range mappings {
+		val, err := extractClaimFromPath(fromClaimPath, claims)
+		if err != nil {
+			log.Debugf("Failed to extract claim from path: %v", err)
+			continue
+		}
+		if err := addClaimToUserClaims(externalUserClaim, toClaimName, val); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func addClaimToUserClaims(externalUserClaim *tokens.ExternalUserClaim, attributeName string, claimValue interface{}) error {
+	switch v := claimValue.(type) {
+	case []interface{}:
+		for i, arrayVal := range v {
+			_, isArray := arrayVal.([]interface{})
+			_, isNestedStruct := arrayVal.(map[string]interface{})
+			if isArray || isNestedStruct {
+				return errors.Errorf("Unsupported claim type %T with value %+v", arrayVal, arrayVal)
+			}
+			if err := addClaimToUserClaims(externalUserClaim, attributeName, arrayVal); err != nil {
+				return errors.Wrapf(err, "failed to add %d element of %+v", i, v)
+			}
+		}
+	case string:
+		externalUserClaim.Attributes[attributeName] = append(externalUserClaim.Attributes[attributeName], v)
+	case bool:
+		externalUserClaim.Attributes[attributeName] = append(externalUserClaim.Attributes[attributeName], strconv.FormatBool(v))
+	default:
+		return errors.Errorf("Unsupported claim type %T with value %+v", claimValue, claimValue)
+	}
+	return nil
+}
+
+func extractClaimFromPath(fromClaimName string, claims map[string]interface{}) (interface{}, error) {
+	claimPath := strings.Split(fromClaimName, ".")
+	currentNode := claims
+	for i, next := range claimPath {
+		nextVal, ok := currentNode[next]
+		if !ok {
+			return nil, errors.Errorf("no value %q on the path %q", next, fromClaimName)
+		}
+		if i == len(claimPath)-1 {
+			return nextVal, nil
+		}
+		currentNode, ok = nextVal.(map[string]interface{})
+		if !ok {
+			return nil, errors.Errorf("expected next value to be of map type but got %T", nextVal)
+		}
+	}
+	return nil, nil
```

Comment:
This looks weird. Is it because we are never supposed to reach this line? If so, I'd suggest to:
* either change the loop control flow to return `nextVal, nil` upon completion
* or add a warning log line between after the loop saying unexpected instructions are called and return an error in L713.